fastcrypto-zkp/src/circom.rs

```diff
@@ -3,20 +3,106 @@
 
 use ark_bn254::{Fq, Fq2, G1Affine, G1Projective, G2Affine, G2Projective};
 use fastcrypto::error::FastCryptoError;
-
+use num_bigint::BigUint;
+use schemars::JsonSchema;
+use serde::{Deserialize, Serialize};
 /// A G1 point in BN254 serialized as a vector of three strings which is the canonical decimal
 /// representation of the projective coordinates in Fq.
-pub type CircomG1 = Vec<String>;
+pub type CircomG1 = Vec<Bn254FieldElement>;
 
 /// A G2 point in BN254 serialized as a vector of three vectors each being a vector of two strings
 /// which are the canonical decimal representation of the coefficients of the projective coordinates
 /// in Fq2.
-pub type CircomG2 = Vec<Vec<String>>;
+pub type CircomG2 = Vec<Vec<Bn254FieldElement>>;
+
+/// A struct that can represent a Bn254 point serialized as a radix10 encoded string.
+#[derive(Debug, Clone, JsonSchema, Eq, PartialEq)]
+pub struct Bn254FieldElement([u8; 32]);
+
+impl Bn254FieldElement {
+    /// Returns the unpadded version of the point. This is the point with leading zeros removed.
+    pub fn unpadded(&self) -> &[u8] {
+        let mut buf = self.0.as_slice();
+
+        while !buf.is_empty() && buf[0] == 0 {
+            buf = &buf[1..];
+        }
+
+        // If the value is '0' then just return a slice of length 1 of the final byte
+        if buf.is_empty() {
+            &self.0[31..]
+        } else {
+            buf
+        }
+    }
+
+    /// Returns the padded version of the point. This is the point with leading zeros preserved to 32 bytes.
+    pub fn padded(&self) -> &[u8] {
+        &self.0
+    }
+}
+impl std::str::FromStr for Bn254FieldElement {
+    type Err = Bn254FieldElementParseError;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        let big_int = Fq::from_str(s).map_err(|_| Bn254FieldElementParseError::InvalidInput)?;
+        let be_bytes = BigUint::from_str(&big_int.to_string())?.to_bytes_be();
```

Comment:
Use `big_int.0.to_byte_be()` instead of the slower `from_str`.

Reply:
This is, btw, guaranteed to be padded to 32 bytes, so the extra length check below is not needed.

Reply:
Something like this should work:
```
        let field_element = Fq::from_str(s).map_err(|_| Bn254FieldElementParseError::InvalidInput)?;
        let be_bytes = field_element.0.to_bytes_be();
        be_bytes.try_into().map_err(|_| Bn254FieldElementParseError::InvalidInput).map(Bn254FieldElement)
```
As mentioned above, the error in `try_into` will never fail, but I think it's alright to handle it anyway.